### PR TITLE
Fix WritableBufferWriter.Write bounds check

### PR DESCRIPTION
**Pull request by benaadams**
Resolves https://github.com/dotnet/corefxlab/issues/1366

/cc @pakrym @davidfowl

### CODE REVIEW

**Comment by benaadams** [comment]
Testing wrong length; needs changes

**Comment by benaadams** [comment]
@pakrym added tests 😉

**Comment by pakrym** [comment]
CI failures

**Comment by benaadams** [comment]
Was a d'oh issue

**Comment by benaadams** [comment]
@pakrym fixed the 0 length things...

File: tests/System.IO.Pipelines.Tests/WritableBufferWriterFacts.cs [hunk 1]
```diff
@@ -59,6 +59,30 @@ public void SlicesSpanAndAdvancesAfterWrite()
...
+            _buffer = _pipe.Writer.Alloc(1);
+            var initialLength = _buffer.Buffer.Length;
+
+            var writer = new WritableBufferWriter(_buffer);
+            var array = new byte[] { 1, 2, 3 };
+
+            writer.Write(array, 0, 0);
+            writer.Write(array, array.Length, 0);
+
+            Assert.Throws<ArgumentOutOfRangeException>(() => writer.Write(array, -1, 0));
```
Comment (ahsonkhan) [inline tests/System.IO.Pipelines.Tests/WritableBufferWriterFacts.cs]: We can't do this with a type that wraps a Span. FYI.

Comment (benaadams) [inline tests/System.IO.Pipelines.Tests/WritableBufferWriterFacts.cs]: Its an array though?

Comment (benaadams) [inline tests/System.IO.Pipelines.Tests/WritableBufferWriterFacts.cs]: Ahh... the writer?

Comment (ahsonkhan) [inline tests/System.IO.Pipelines.Tests/WritableBufferWriterFacts.cs]: Fixed it in here:
https://github.com/dotnet/corefxlab/pull/1401/commits/d52726144b5f94dfcba9e2302555a4b64806c00c

**Pull request merged**